### PR TITLE
Call ResetTimer synchronously from eol and skip the Tick when channel is null

### PR DESCRIPTION
**Pull request by Pldi23**
Cloudbees CI reported a flakiness in TimeoutStepExecutionTest#activityRestart.

Logs:
[activityRestart_stacktrace.txt](https://github.com/jenkinsci/workflow-basic-steps-plugin/files/9297931/activityRestart_stacktrace.txt)
[activityRestart_stderr.txt](https://github.com/jenkinsci/workflow-basic-steps-plugin/files/9297933/activityRestart_stderr.txt)

The issue could be reproduced locally, (test crashes quite often), when after restart timeout set to expire in less then 7.5 sec the test fails.
```
1.281 [restarted #1] Timeout set to expire after 3.3 sec without activity
```

From what I see the reason of the flake is the race condition between Tick (which gets scheduled with timeout / 2, in our particular case 7.5 sec) and Killer's delay (which is after restart in our particular case was 3.3 sec). When `delay < timeout / 2` no reset happens and TimeoutStepExecution cancelled the rest of the pipeline after delay. 

There is a `TODO if channel == null, we can safely ResetTimer.call synchronously from eol and skip the Tick`, if we will follow this recommendation we could reset timer synchronously instead of scheduling it in `timeout / 2` time, which in itself will fix a flakiness.

### CODE REVIEW

**Comment by Pldi23** [comment]
@jglick Could you please review? I tried to apply a with a minimal impact to not break anything.

File: src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepExecution.java [hunk 1]
```diff
@@ -336,11 +336,13 @@ private Object readResolve() {
         @Override
         public OutputStream decorateLogger(@SuppressWarnings("rawtypes") Run build, final OutputStream logger)
                 throws IOException, InterruptedException {
-            // TODO if channel == null, we can safely ResetTimer.call synchronously from eol and skip the Tick
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepExecution.java:339]: was from #74 FTR

**Review by jglick — APPROVED** [review]
OK I guess? I have no memory of writing that comment. Does it seem to work?

**Comment by Pldi23** [comment]
> OK I guess? I have no memory of writing that comment. Does it seem to work?

Flakiness gone with this change. All other tests works. Just in case I also tested manually the case described in https://issues.jenkins.io/browse/JENKINS-54078?focusedCommentId=351432&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-351432. So yes, It seem to work.

**Comment by chwehrli** [comment]
After upgrading to v991 (which includes this PR) we're now facing cpu pressure and eventually Jenkins POD is being terminated.
I believe this might be caused by some of our jobs producing massive logs (>50k lines) within minutes and hence now triggering lots of ResetTimer calls:

`	org.jenkinsci.plugins.workflow.steps.TimeoutStepExecution$ResetTimer.call(TimeoutStepExecution.java:308)
	org.jenkinsci.plugins.workflow.steps.TimeoutStepExecution$ConsoleLogFilterImpl2$1.eol(TimeoutStepExecution.java:344)
	hudson.console.LineTransformationOutputStream.eol(LineTransformationOutputStream.java:61)
`

[JavaMelody__jenkins-0_8_29_22.pdf](https://github.com/jenkinsci/workflow-basic-steps-plugin/files/9446033/JavaMelody__jenkins-0_8_29_22.pdf)

[threadDump.txt](https://github.com/jenkinsci/workflow-basic-steps-plugin/files/9446034/threadDump.txt)

Any suggestions?
Please let me know if you want me to report a bug for this issue.

Many thanks!
Christian

**Comment by tarioch** [comment]
Maybe worth adding, the jobs run with a timeout of 2h, so if I got this right the reset call before happened only every 1h and would now happen on each log line.

**Comment by jglick** [comment]
> The issue could be reproduced locally, (test crashes quite often), when after restart timeout set to expire in less then 7.5 sec the test fails.

via

```diff
diff --git src/test/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepTest.java src/test/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepTest.java
index f18f476..faea2f4 100644
--- src/test/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepTest.java
+++ src/test/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepTest.java
@@ -219,6 +219,7 @@ public class TimeoutStepTest {
                 WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
                 SemaphoreStep.waitForStart("restarted/1", b);
         });
+        Thread.sleep(10_000);
         sessions.then(j -> {
                 WorkflowJob p = j.jenkins.getItemByFullName("restarted", WorkflowJob.class);
                 WorkflowRun b = p.getBuildByNumber(1);
```

**Comment by jglick** [comment]
@chwehrli @tarioch thanks for the report and sorry for any inconvenience. I believe #234 should fix the performance issue as well as address the original problem motivating this PR. If you want to test prior to release, that would be great; just download a `workflow-basic-steps.hpi` from the **Incrementals** link of #234, assuming tests pass.

**Comment by chwehrli** [comment]
Thank you very much for that ultra-fast response @jglick !
I've just installed workflow-basic-steps-992.v85b_6c33ca_6e0.hpi - will report later how it behaves...

**Comment by jglick** [comment]
Thanks. Please use #234 for further comments.

**Pull request merged**